lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.cpp

```diff
@@ -0,0 +1,300 @@
+#include "lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <iomanip>
+#include <ios>
+#include <numeric>
+#include <sstream>
+#include <string>
+
+namespace mlir {
+namespace heir {
+namespace bgv {
+
+// the formulae below are mainly taken from MMLGA22
+// "Finding and Evaluating Parameters for BGV"
+// https://eprint.iacr.org/2022/706
+
+template <bool P>
+using Model = NoiseByVarianceCanEmbModel<P>;
+
+template <bool P>
+double Model<P>::toLogBound(const LocalParamType &param,
+                            const StateType &noise) {
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  // StateType only stores e in (m + t * e), so when we want to print the bound
+  // we need to multiply t back
+  return log(t * noise.getValue()) / log(2);
+}
+
+template <bool P>
+double Model<P>::toLogBudget(const LocalParamType &param,
+                             const StateType &noise) {
+  return toLogTotal(param) - toLogBound(param, noise);
+}
+
+template <bool P>
+double Model<P>::toLogTotal(const LocalParamType &param) {
+  double total = 0;
+  auto logqi = param.getSchemeParam()->getLogqi();
+  for (auto i = 0; i <= param.getCurrentLevel(); ++i) {
+    total += logqi[i];
+  }
+  return total - 1.0;
+}
+
+template <bool P>
+std::string Model<P>::toLogBoundString(const LocalParamType &param,
+                                       const StateType &noise) {
+  auto logBound = toLogBound(param, noise);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logBound;
+  return stream.str();
+}
+
+template <bool P>
+std::string Model<P>::toLogBudgetString(const LocalParamType &param,
+                                        const StateType &noise) {
+  auto logBudget = toLogBudget(param, noise);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logBudget;
+  return stream.str();
+}
+
+template <bool P>
+std::string Model<P>::toLogTotalString(const LocalParamType &param) {
+  auto logTotal = toLogTotal(param);
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(2) << logTotal;
+  return stream.str();
+}
+
+template <bool P>
+double Model<P>::getExpansionFactor(const LocalParamType &param) {
+  auto phi = param.getSchemeParam()->getRingDim() / 2;
+  return phi;
+}
+
+template <bool P>
+double Model<P>::getBoundErr(const LocalParamType &param) {
+  auto std0 = param.getSchemeParam()->getStd0();
+  // probability of larger than 6 * std0 is less than 2^{-30}
+  auto assurance = 6;
```

Comment:
Assurance here is the same term as FactorD below. Maybe we can merge them. I get the name assurance from Openfhe codebase.

But as in canonical embedding we also do not need `BoundErr`, we can also delete this function safely. `BoundKey` below can also be deleted.

Reply:
Deleted